api/v1beta1/types.go

```diff
@@ -84,9 +84,7 @@ type NetworkSpec struct {
 	// +optional
 	ControlPlaneOutboundLB *LoadBalancerSpec `json:"controlPlaneOutboundLB,omitempty"`
 
-	// PrivateDNSZoneName defines the zone name for the Azure Private DNS.
-	// +optional
-	PrivateDNSZoneName string `json:"privateDNSZoneName,omitempty"`
+	NetworkCommon `json:",inline"`
```

Comment:
A generic naming question: I'm curious if the `Common` suffix is a pattern that other repos/projects dealing with similar situations with CRDs and Templates have adopted, or if it's something new that we're introducing here? If it's the later, have you considered how others are naming types and if there are any existing conventions out there?

Reply:
The closest thing I could find similar to this was `ObjectMeta` used in all types. This is different to the use case here, as we are trying to group a subset of fields into another type. As to the choice  of `Common` prefix, I couldn't really think of anything better (perhaps `Shared`) although I'm happy to consider other suggestions.

Reply:
I asked about this during today's CAPI office hours and here are the pointers I got: 

[stefan] We had a similar discussion for our ClusterClass structs and I did some research. There doesn’t seem to be a prefix/suffix pattern in Core Kubernetes types for shared structs, but some specific structs are prefixed (e.g. EndpointPort, EndpointAddress,... ([link](https://github.com/kubernetes-sigs/cluster-api/pull/5418#issuecomment-947893113))).
[vince] Closest example pattern can be found in https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/examples/guestbook-operator/api/v1alpha1/guestbook_types.go
[yassine] what we do in CAPV:  https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/apis/v1beta1/vspheremachine_types.go#L34 and https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/apis/v1beta1/vspherevm_types.go#L34


https://docs.google.com/document/d/1ushaVqAKYnZ2VN_aa3GyKlS4kEd6bSug13xaXOakAQI/edit#

Reply:
Thanks Cecile for bringing this up in capi meeting, apologies couldn't attend it today. From the examples above, I feel the [GuestBook](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/master/examples/guestbook-operator/api/v1alpha1/guestbook_types.go) closely resembles to what we are trying to do here. wdyt about `CommonAzureClusterSpec`? 

[capv](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/master/apis/v1beta1/vspheremachine_types.go#L34) example also looks interesting in that we are also extracting information for cloning AzureCluster templates, but I'm not entirely sure how it'll look for other composed fields (eg. `NetworkCloneSpec`?)

Reply:
> From the examples above, I feel the GuestBook closely resembles to what we are trying to do here. wdyt about CommonAzureClusterSpec?

I agree, +1 to CommonAzureClusterSpec

Reply:
Holding this change for more feedback. 
cc @devigned

Reply:
@devigned @sbueringer @fabriziopandini @killianmuldoon do you have thoughts on this one?

current proposed naming is `Common` prefix, eg. `CommonAzureClusterSpec` for the struct that holds fields common between AzureCluster Spec and AzureClusterTemplate Spec. Trying to see if this would work as a pattern if cluster api ends up doing something similar to reuse parts of specs for template types for ClusterClass.

Reply:
Hm good question. In general I think that:
1. types very specific to a certain API type in a way that they cannot be used later in other resources should be prefixed  with the API type
2. generic types ideally shouldn't have any prefix

Examples:
1. KubeadmControlPlaneSpec, KubeadmControlPlaneStatus, EndpointPort, ..
2. clusterv1.Conditions, APIEndpoint, ClusterVariable, JSONPatch

But not sure if that would be a good convention, given that it's hard to forecast future usages of structs. Also if every API type specific struct would be prefixed it can lead to unreadable struct names in deeply nested structs.

I think this is is also a special case. In general I would say that's the 2nd case (although it's not very generic, just used in two places), so I wouldn't use a prefix. But the name without prefix `AzureClusterSpec` already includes the API type `AzureCluster` as prefix which could be confusing as the struct is also used in `AzureClusterTemplate`, but maybe that's not too problematic.

Probably the most important thing should be readability of the struct name, but I think both `CommonAzureClusterSpec` and `AzureClusterSpec` would be fine. 

tl;dr Sorry this probably doesn't help at all. I think in general I'm fine with all options, I'm just not sure if we want to prefix every struct which is used in multiple API types or could be used in multiple types in the future with "Common".

Reply:
I tried to avoid the "Common" prefix/suffix with something more meaningful to what the type represents, but couldn't find any. In this case, we want to name two types which represents the following:

1. `Type 1` - Fields that could be shared by multiple azure clusters. - `CommonAzureClusterSpec`
2. `Type 2` - Fields that are unique to each azure cluster - `AzureClusterSpec`

Another option is, we can call `Type 1` as `AzureClusterSpec`, and rename `Type 2` as `AzureClusterInstanceSpec` to represent one instance of AzureCluster whereas `AzureClusterTemplate` would represent a class of AzureClusters?

Just to recap, the other option, if we wanted to avoid this refactoring, would be to duplicate the specs for AzureClusterTemplate, which we didn't want to pursue because we'd then have to maintain two sets of fields. One way to workaround this would be to write a trivial type comparator that would throw warnings if there are any changes made to those types. Kind of something similar to api-diff.

Reply:
What about `AzureClusterClassSpec`?

Reply:
> What about `AzureClusterClassSpec` and apply the same to other structures. It's kind of the inverse of `AzureClusterInstanceSpec`.

+1, just to confirm for nested fields, it will be suffixed with `ClassSpec` and not `ClusterClassSpec` right? Eg. `NetworkClassSpec`. wdyt @CecileRobertMichon

Reply:
+1 to `ClassSpec` suffix